cmd/openshift-compatibility-gen/generators/gen_type.go

```diff
@@ -0,0 +1,226 @@
+package generators
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	"k8s.io/gengo/generator"
+	"k8s.io/gengo/namer"
+	"k8s.io/gengo/types"
+	"k8s.io/klog/v2"
+)
+
+// NewCompatibilityGen creates a generator for the prerelease-lifecycle-generator
+func NewCompatibilityGen(sanitizedName, targetPackage string) generator.Generator {
+	return &genCompatibility{
+		DefaultGen: generator.DefaultGen{
+			OptionalName: sanitizedName,
+		},
+		targetPackage: targetPackage,
+		imports:       generator.NewImportTracker(),
+		typesForInit:  make([]*types.Type, 0),
+	}
+}
+
+// genCompatibility produces a file with autogenerated deep-copy functions.
+type genCompatibility struct {
+	generator.DefaultGen
+	targetPackage string
+	imports       namer.ImportTracker
+	typesForInit  []*types.Type
+}
+
+func (g *genCompatibility) Namers(c *generator.Context) namer.NameSystems {
+	return namer.NameSystems{
+		"public":       namer.NewPublicNamer(1),
+		"intrapackage": namer.NewPublicNamer(0),
+		"raw":          namer.NewRawNamer("", nil),
+	}
+}
+
+func (g *genCompatibility) Filter(c *generator.Context, t *types.Type) bool {
+	// Filter out types not being processed or not copyable within the package.
+	if !isAPIType(t) {
+		klog.V(2).Infof("Type %v is not a valid target for compatibility", t)
+		return false
+	}
+	g.typesForInit = append(g.typesForInit, t)
+	return true
+}
+
+func (g *genCompatibility) Imports(c *generator.Context) (imports []string) {
+	importLines := []string{}
+	for _, singleImport := range g.imports.ImportLines() {
+		if g.isOtherPackage(singleImport) {
+			importLines = append(importLines, singleImport)
+		}
+	}
+	return importLines
+}
+
+func (g *genCompatibility) isOtherPackage(pkg string) bool {
+	if pkg == g.targetPackage {
+		return false
+	}
+	if strings.HasSuffix(pkg, "\""+g.targetPackage+"\"") {
+		return false
+	}
+	return true
+}
+
+func (g *genCompatibility) Init(c *generator.Context, w io.Writer) error {
+	return nil
+}
+
+func (g *genCompatibility) GenerateType(c *generator.Context, t *types.Type, w io.Writer) error {
+	klog.V(3).Infof("Generating prerelease-lifecycle for type %v", t)
+
+	sw := generator.NewSnippetWriter(w, c, "$", "$")
+	args, err := g.argsFromType(c, t)
+	if err != nil {
+		return err
+	}
+
+	if versionedMethodOrDie("OpenShiftAPICompatibilityLevel", t) == nil {
```

Comment:
I think you have a package name in here for this type

1. If you cannot parse out a `vN{,betaM,alphaM}`, then produce non-compiling code indicating that
2. if alpha and compatibility level not equal four, then produce non-compiling code explaining why and linking doc
3. if beta compatibility level is not two, this is not allowed.  If they choose one, they should be GA. If they choose less, they should be alpha,
4. if GA compatibility level is not one, this is not allowed.  GA APIs are supported for 12 months

Reply:
Done.

Reply:
> Done.

I bet that you get more bugs about a failed generator than you do about non-compiling code (which they will check into their branhc), but I'm willing to try.